### PR TITLE
Qc on speed: FastP and Mosdepth as drop-in replacement for Trimgalore/Seqkit & Deeptools/qualimap

### LINKED ISSUE #508: [FEATURE] Investigate Fastp usage for speed improvement
cc @marchoeppner

### PR DESCRIPTION
**Pull request by FriederikeHanssen**
Close #508 

cc #95 

TODO:

- [x] Update Fastp module to not have .trim in the output (confusing when it is just used for splitting
- [x]  UMI subworkflow: incooporation into FastP?
- [x] Add test for trimming
- [x] Update subway map
- [x] Update docs
- [x] Make sure MultiQC has all plots
- [x] Document new parameters, update modules.config

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/modules.config [hunk 1]
```diff
@@ -193,23 +193,33 @@ process {
...
+            params.trim_fastq              ?: "--disable_adapter_trimming",
+            params.clip_r1 > 0             ? "--trim_front1 ${params.clip_r1}" : "", // Remove bp from the 5' end of read 1.
+            params.three_prime_clip_r1 > 0 ? "--trim_tail1 ${params.three_prime_clip_r1}" : "", // Remove bp from the 3' end of read 1 AFTER adapter/quality trimming has been performed.
+            params.clip_r2   > 0           ? "--trim_front2 ${params.clip_r2}" : "", //Remove bp from the 5' end of read 2.
+            params.three_prime_clip_r2 > 0 ? "--trim_tail2 ${params.three_prime_clip_r2}" : "", // Remove bp from the 3' end of read 2 AFTER adapter/quality trimming has been performed.
+            params.trim_nextseq            ? "--trim_poly_g" : "", // Apply the --nextseq=X option, to trim based on quality after removing poly-G tails.
+            params.split_fastq > 0         ? "--split_by_lines ${params.split_fastq * 4}" : "",
+
+            // For trimgalore also auto detection was used
+            //params.adapter_R1       ? "--adapter_sequence ${params.adapter_R1}" : "", //normally not needed
```
Comment (FriederikeHanssen) [inline conf/modules.config:208]: should we provide the additonal parameter for setting specific adapter sequences? Wasn't supported before, but trimgalore guessed the adapter (same here at the moment by default) @maxulysse

File: conf/modules.config [hunk 2]
```diff
@@ -193,23 +193,29 @@ process {
...
-    withName: 'TRIMGALORE' {
-        ext.args         = '--fastqc'
-        publishDir       = [
-            [
-                path: { "${params.outdir}/reports/trimgalore/${meta.id}/fastqc" },
-                mode: params.publish_dir_mode,
-                pattern: "*{html,zip}"
-            ],
+
+    withName: FASTP {
```
Comment (ggabernet) [inline conf/modules.config:197]:
```suggestion
    withName: 'FASTP' {
```

Comment (ggabernet) [reply conf/modules.config:197]: maybe it's not needed, not sure now, but just in case...

File: tests/test_recalibrate.yml [hunk 3]
```diff
@@ -52,7 +58,13 @@
       should_exist: false
     - path: results/csv/recalibrated.csv
       should_exist: false
-    - path: results/reports/qualimap/test/test.recal
+    - path: results/reports/mosdepth/test/test.recal.mosdepth.global.dist.txt
+      should_exist: false
```
Comment (ggabernet) [inline tests/test_recalibrate.yml:62]: Just a question, so is here testing that the files are not existing? Or what is the `should_exist` doing?

File: workflows/sarek.nf [hunk 4]
```diff
@@ -294,13 +291,17 @@ workflow SAREK {
     PREPARE_INTERVALS(fasta_fai)
 
     // Intervals for speed up preprocessing/variant calling by spread/gather
-    intervals_bed_combined        = (params.intervals && params.wes) ? Channel.fromPath(params.intervals).collect() : []
+    // this is not good, we need the combined bed for some tools that don't support scatter/gather. Why would we not use the same intervals for WGS?
+    // intervals_bed_combined        = (params.intervals && params.wes) ? Channel.fromPath(params.intervals).collect() : []
```
Comment (ggabernet) [inline workflows/sarek.nf:295]: If your alternative is working, can these comments then be removed?

File: workflows/sarek.nf [hunk 5]
```diff
@@ -364,25 +350,35 @@ workflow SAREK {
...
-            ch_input_sample_to_split = ch_reads
+            ch_reads_fastp = ch_input_fastq
         }
 
-        // SPLIT OF FASTQ FILES WITH SEQKIT_SPLIT2
-        if (params.split_fastq > 1) {
-            SPLIT_FASTQ(ch_input_sample_to_split)
+        // Trimming and/or splitting
+        if (params.trim_fastq || params.split_fastq > 0) {
+            FASTP(ch_reads_fastp, false, false)
```
Comment (ggabernet) [inline workflows/sarek.nf:364]: How do you specify here if trimming or not? As I read it here it's always trimming reads, even if `trim_fastq` is false, provided that `split_fastq>0`. Is that intended?

**Review by ggabernet — COMMENTED** [review]
I had some comments, but otherwise looks good!

File: workflows/sarek.nf (hunk 5 above)
Comment (FriederikeHanssen) [reply workflows/sarek.nf:364]: In the modules.config it sets:

```ext.args = [ "",
            params.trim_fastq              ?: "--disable_adapter_trimming",
            params.clip_r1 > 0             ? "--trim_front1 ${params.clip_r1}" : "", // Remove bp from the 5' end of read 1.
            params.three_prime_clip_r1 > 0 ? "--trim_tail1 ${params.three_prime_clip_r1}" : "", // Remove bp from the 3' end of read 1 AFTER adapter/quality trimming has been performed.
            params.clip_r2   > 0           ? "--trim_front2 ${params.clip_r2}" : "", //Remove bp from the 5' end of read 2.
            params.three_prime_clip_r2 > 0 ? "--trim_tail2 ${params.three_prime_clip_r2}" : "", // Remove bp from the 3' end of read 2 AFTER adapter/quality trimming has been performed.
            params.trim_nextseq            ? "--trim_poly_g" : "", // Apply the --nextseq=X option, to trim based on quality after removing poly-G tails.
            params.split_fastq > 0         ? "--split_by_lines ${params.split_fastq * 4}" : "",

        ].join(" ").trim()```
        
        so if not params.trim_fastq, aka trimming false, then the tool sets `--disable_adapter_trimming` and shouldn't trim any adapters

**Review by ggabernet — APPROVED** [review]
Looks good to me!

File: workflows/sarek.nf (hunk 4 above)
Comment (FriederikeHanssen) [reply workflows/sarek.nf:295]: oh I almost missed this, no and yes, this warrants a separate PR, that i have already discussed with Oskar also in regard to haplotypecaller. I'll open a card

**Pull request merged**